opal/corelib/array.rb

```diff
@@ -869,12 +869,11 @@ def drop(number)
 
   def dup
     %x{
-      if (
-        self.$$class === Opal.Array &&
-        self.$$class.$allocate.$$pristine &&
-        self.$copy_instance_variables.$$pristine &&
-        self.$initialize_dup.$$pristine
-      ) return self.slice(0);
+      if (self.$$class.$allocate.$$pristine &&
```

Comment:
`self.$$class === Opal.Array` should stay because slice produces an array:
``` ruby
>> class A < Array; end; `#{A.new}.slice().$class()`
=> Array
>> class A < Array; end; A.new.dup.class
=> A
```

So the method itself doesn't need any changes.
Sorry, missed that initially 😄

Reply:
Wait, different classes have different allocators, ignore my previous comment please, everything is OK. There's even a test in the rubyspec's suite that passes.